runtime/contract_update_validation.go

```diff
@@ -0,0 +1,434 @@
+	}()
+
+	// If the same decl is already visited, then do not check again.
+	// This also avoids getting stuck on circular dependencies between composite decls.
+	if validator.visited[validator.currentDeclName] {
+		return nil
+	}
+
+	validator.visited[validator.currentDeclName] = true
+
+	oldFields := getDeclarationMembers(oldDeclaration).Fields()
+	newFields := getDeclarationMembers(newDeclaration).Fields()
+
+	// Updated contract has to have at-most the same number of field as the old contract.
+	// Any additional field may cause crashes/garbage-values when deserializing the
+	// already-stored data. However, having less number of fields is fine for now. It will
+	// only leave some unused-data, and will not do any harm to the programs that are running.
+
+	// This is a fail-fast check.
+	if len(oldFields) < len(newFields) {
+		if validator.isInNestedDecl() {
+			return fmt.Errorf("too many fields in `%s`. expected %d, found %d",
+				validator.currentDeclName, len(oldFields), len(newFields))
+		}
+
+		return fmt.Errorf("too many fields. expected %d, found %d",
+			len(oldFields), len(newFields))
+	}
+
+	// Put the old field-types against their field-names to a map, for faster lookup.
+	oldFiledTypes := map[string]ast.Type{}
+	for _, field := range oldFields {
+		oldFiledTypes[field.Identifier.Identifier] = field.TypeAnnotation.Type
+	}
+
+	for _, newField := range newFields {
+		err := validator.visitField(newField, oldFiledTypes)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) visitField(
+	newField *ast.FieldDeclaration,
+	oldFiledTypes map[string]ast.Type) error {
+
+	prevFieldName := validator.currentFieldName
+	validator.currentFieldName = newField.Identifier.Identifier
+	defer func() {
+		validator.currentFieldName = prevFieldName
+	}()
+
+	oldType := oldFiledTypes[validator.currentFieldName]
+	if oldType == nil {
+		if validator.isInNestedDecl() {
+			return fmt.Errorf("found new field `%s` in `%s",
+				validator.currentDeclName,
+				validator.currentFieldName)
+		}
+
+		return fmt.Errorf("found new field `%s`", validator.currentFieldName)
+	}
+
+	return oldType.Equal(newField.TypeAnnotation.Type, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckNominalTypeEquality(expected *ast.NominalType, found ast.Type) error {
+	foundNominalType, ok := found.(*ast.NominalType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// First check whether the names are equal.
+	ok = validator.checkNameEquality(expected, foundNominalType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// Then, check whether the fields of the two composite declarations
+	// referred by this nominal type, are also compatible.
+
+	var compositeDeclName string
+	if isQualifiedName(foundNominalType) {
+		compositeDeclName = foundNominalType.NestedIdentifiers[0].Identifier
+	} else {
+		compositeDeclName = foundNominalType.Identifier.Identifier
+	}
```

Comment:
This check assumes that nesting will always be one level deep. Maybe we can generalize this in case we allow other nesting of declarations in the future?

Reply:
this snippet is no longer used now.